clojure-lsp: migrate to GraalVM

thiagokokada opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

clojure-lsp now supports GraalVM builds that makes the binary faster to start and also seems to reduce memory usage.

Also bump version, to ensure we are using the latest fixes for GraalVM builds.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

thiagokokada committed:
clojure-lsp: migrate to GraalVM

clojure-lsp now supports GraalVM builds that makes the binary faster to
start and also seems to reduce memory usage.

thiagokokada commented on pkgs/development/tools/misc/clojure-lsp/default.nix:
```diff
@@ -1,31 +1,63 @@
+          "-J-Dclojure.spec.skip-macros=true"
+          "-H:+ReportExceptionStackTraces"
+          "--enable-url-protocols=jar"
+          # "-H:+InlineBeforeAnalysis"
```
To be enabled in GraalVM 21.0.0.

thiagokokada commented on pkgs/development/tools/misc/clojure-lsp/default.nix:
```diff
@@ -1,31 +1,63 @@
+  buildPhase = with lib; ''
+    args=("-jar" "${src}"
+          "-H:Name=clojure-lsp"
+          ${optionalString stdenv.isDarwin ''"-H:-CheckToolchain"''}
```
Didn't test, just copied it from `babashka` derivation.

SuperSandro2000 replied:
Will I notice something different when building on darwin?

thiagokokada replied:
Don't know, I don't have a macOS to test.

thiagokokada commented:
@GrahamcOfBorg build clojure-lsp

thiagokokada committed:
clojure-lsp: add tests

thiagokokada commented:
Converting it in draft since the tests doesn't pass.

@GrahamcOfBorg build clojure-lsp

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 111864` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and are new build failure:</summary>
  <ul>
<li>clojure-lsp: log <a href="https://termbin.com/chog">https://termbin.com/chog</a></li>
  </ul>
</details>

ericdallo reviewed (approved):
LGTM, just need to check why tests are not passing

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 111864` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and are new build failure:</summary>
  <ul>
<li>clojure-lsp: log <a href="https://termbin.com/kxb1">https://termbin.com/kxb1</a></li>
  </ul>
</details>

ericdallo commented:
Ok, the issue with the tests is that we are not using the same jar clojure-lsp use to build the native-image, I'll change the clojure-lsp release to provide that jar too

thiagokokada commented:
@GrahamcOfBorg build clojure-lsp

thiagokokada commented:
@GrahamcOfBorg build clojure-lsp

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 111864` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and are new build failure:</summary>
  <ul>
<li>clojure-lsp: log <a href="https://termbin.com/rqm4">https://termbin.com/rqm4</a></li>
  </ul>
</details>

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 111864` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and are new build failure:</summary>
  <ul>
<li>clojure-lsp: log <a href="https://termbin.com/0ayw">https://termbin.com/0ayw</a></li>
  </ul>
</details>

thiagokokada commented:
Seems we will probably need to bump GraalVM to version 21.0.0 first.

thiagokokada committed:
clojure-lsp: build it from source

Nowadays we build it from the jar released from upstream, but this PRs
changes it to build it from source code.

thiagokokada commented:
Commit f6490f2 is building clojure-lsp completely from source. It will not work since `clojure-lsp` needs GraalVM 21.0.0 to compile, but since it is not working right now I decided to push this refactoring.

thiagokokada commented:
Should be ready to merge now.

@SuperSandro2000 @ericdallo, can you please review?

thiagokokada commented on pkgs/development/tools/misc/clojure-lsp/default.nix:
```diff
@@ -1,31 +1,55 @@
+    install -Dm755 ./clojure-lsp $out/bin/clojure-lsp
+  '';
+
+  # TODO: tests are timing out, needs more investigation
+  doCheck = false;
+  checkPhase = ''
+    ${babashka}/bin/bb ${src}/integration-test/run-all.clj ./clojure-lsp
   '';
```
Running those tests manually seems to work, so probably there is some other issue.

thiagokokada commented:
@GrahamcOfBorg build clojure-lsp

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 111864` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and already failed to build on hydra master:</summary>
  <ul>
<li>clojure-lsp: <a href="https://termbin.com/cxua">plain log</a>  |  <a href="https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/"content" wasn't supplied./raw/index.html">pretty log </a></li>
  </ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
clojure-lsp:
<details>


`buildPhase` should probably contain `runHook preBuild` and `runHook postBuild`.

Near pkgs/development/tools/misc/clojure-lsp/default.nix:33:3:

```
   |
33 |   buildPhase = with lib; ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
`checkPhase` should probably contain `runHook preCheck` and `runHook postCheck`.

Near pkgs/development/tools/misc/clojure-lsp/default.nix:44:3:

```
   |
44 |   checkPhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/development/tools/misc/clojure-lsp/default.nix:38:3:

```
   |
38 |   installPhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
Unused argument: `fetchurl`.
Near pkgs/development/tools/misc/clojure-lsp/default.nix:1:29:

```
  |
1 | { lib, stdenv, callPackage, fetchurl, fetchFromGitHub, leiningen, openjdk11
  |                             ^
```
</details>
</details>

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 111864` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>clojure-lsp</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
clojure-lsp:
<details>


`buildPhase` should probably contain `runHook preBuild` and `runHook postBuild`.

Near pkgs/development/tools/misc/clojure-lsp/default.nix:33:3:

```
   |
33 |   buildPhase = with lib; ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
`checkPhase` should probably contain `runHook preCheck` and `runHook postCheck`.

Near pkgs/development/tools/misc/clojure-lsp/default.nix:44:3:

```
   |
44 |   checkPhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/development/tools/misc/clojure-lsp/default.nix:38:3:

```
   |
38 |   installPhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
Unused argument: `fetchurl`.
Near pkgs/development/tools/misc/clojure-lsp/default.nix:1:29:

```
  |
1 | { lib, stdenv, callPackage, fetchurl, fetchFromGitHub, leiningen, openjdk11
  |                             ^
```
</details>
</details>

thiagokokada commented:
Fixed the issue with the tests, now it is working.

@ericdallo Can you review it please?

thiagokokada commented:
@GrahamcOfBorg build clojure-lsp
@GrahamcOfBorg test clojure-lsp

ericdallo reviewed (approved):
Nice, thank you very much!

thiagokokada commented:
@SuperSandro2000 Can we have this PR merged?

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 111864` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>clojure-lsp</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
clojure-lsp:
<details>


Unused argument: `fetchurl`.
Near pkgs/development/tools/misc/clojure-lsp/default.nix:1:29:

```
  |
1 | { lib, stdenv, callPackage, fetchurl, fetchFromGitHub, leiningen, openjdk11
  |                             ^
```
</details>
</details>

SuperSandro2000 commented on pkgs/development/tools/misc/clojure-lsp/default.nix:
```diff
@@ -1,31 +1,66 @@
-{ lib, stdenv, fetchurl, jre, makeWrapper }:
+{ lib, stdenv, callPackage, fetchurl, fetchFromGitHub, leiningen, openjdk11
```
```suggestion
{ lib, stdenv, callPackage, fetchFromGitHub, leiningen, openjdk11
```

thiagokokada commented:
One last rebase since @r-ryantm created another bump PR.

Could we approve and merge this? @SuperSandro2000 @ericdallo ?
